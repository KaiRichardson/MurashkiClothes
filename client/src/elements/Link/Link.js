import React from 'react';
import styled, { css } from 'styled-components';
import { Link as L } from 'react-router-dom';
import PropTypes from 'prop-types';

import { black, red, transition, lightGrey } from 'utils';

export const Link = styled(({ dark, light, ...rest }) => <L {...rest} />)`
  --txtColor: ${black};

  ${transition({ prop: 'color' })};
  text-decoration: none;

  /* Default Link */
  color: var(--txtColor);

  &:hover {
    --txtColor: ${red};
  }

  /* Dark Link */
  ${props =>
    props.dark &&
    css`
      --txtColor: ${lightGrey};

      &:hover {
        --txtColor: ${black};
      }
    `}

  /* Light Link */
  ${props =>
    props.light &&
    css`
      --txtColor: ${lightGrey};

      &:hover {
        --txtColor: ${red};
      }
    `}
`;

Link.propTypes = {
  dark: PropTypes.bool
};
