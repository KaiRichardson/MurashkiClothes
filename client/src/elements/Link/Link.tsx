import React from 'react';
import styled, { css } from 'styled-components';
import { Link as L } from 'react-router-dom';

import { black, red, transition, lightGrey } from 'utils';

// eslint-disable-next-line
export const Link = styled(({ dark, light, ext, ...rest }) => (ext ? <a {...rest} /> : <L {...rest} />))<{
  dark?: boolean;
  light?: boolean;
  ext?: boolean;
}>`
  --txtColor: ${black};

  ${transition({ prop: 'color' })};
  text-decoration: none;

  /* Default Link */
  color: var(--txtColor);

  &:hover {
    --txtColor: ${red};
  }

  /* Dark Link */
  ${props =>
    props.dark &&
    css`
      --txtColor: ${lightGrey};

      &:hover {
        --txtColor: ${black};
      }
    `}

  /* Light Link */
  ${props =>
    props.light &&
    css`
      --txtColor: ${lightGrey};

      &:hover {
        --txtColor: ${red};
      }
    `}
`;
