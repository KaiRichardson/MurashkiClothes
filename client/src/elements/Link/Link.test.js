import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import { Link } from './Link';

describe('<Link /> tests', () => {
  it('renders a link', () => {
    //* Act
    const { getByTestId } = render(
      <Router>
        <Link data-testid='test-link' to='/test'>
          test link
        </Link>
      </Router>
    );

    //* Assert
    const testLink = getByTestId('test-link');

    expect(testLink.getAttribute('href')).toBe('/test');

    // Snapshot test
    expect(testLink).toMatchSnapshot();

    // Color test
    expect(testLink).toHaveStyleRule('--txtColor', '#000');
    expect(testLink).toHaveStyleRule('color', 'var(--txtColor)');

    // Hover test
    expect(testLink).toHaveStyleRule('--txtColor', '#ee4266', {
      modifier: ':hover'
    });
  });
});
