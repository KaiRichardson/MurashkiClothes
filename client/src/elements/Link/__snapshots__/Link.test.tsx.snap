// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> tests renders a dark link 1`] = `
.c0 {
  --txtColor: #000;
  -webkit-transition: color 250ms ease-in;
  transition: color 250ms ease-in;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: var(--txtColor);
  --txtColor: #b3b3b3;
}

.c0:hover {
  --txtColor: #ee4266;
}

.c0:hover {
  --txtColor: #000;
}

<a
  class="c0"
  data-testid="test-link"
  href="/test"
>
  test link
</a>
`;

exports[`<Link /> tests renders a default link 1`] = `
.c0 {
  --txtColor: #000;
  -webkit-transition: color 250ms ease-in;
  transition: color 250ms ease-in;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: var(--txtColor);
}

.c0:hover {
  --txtColor: #ee4266;
}

<a
  class="c0"
  data-testid="test-link"
  href="/test"
>
  test link
</a>
`;

exports[`<Link /> tests renders a light link 1`] = `
.c0 {
  --txtColor: #000;
  -webkit-transition: color 250ms ease-in;
  transition: color 250ms ease-in;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: var(--txtColor);
  --txtColor: #b3b3b3;
}

.c0:hover {
  --txtColor: #ee4266;
}

.c0:hover {
  --txtColor: #ee4266;
}

<a
  class="c0"
  data-testid="test-link"
  href="/test"
>
  test link
</a>
`;
